rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,554 @@
+outside of the appender, without any significant upsides identified.
+
+3. centralize the logging system in the coordinator
+
+We could go further, and change the way the logging system works in clustering mode by having the coordinator centralize 
+the logging system. The worker’s logger implementation would just send messages to the coordinator. If this may be a 
+correct design, the main downside is that the logging implementation would be totally different in cluster and 
+non cluster mode, and seems to be way more work that the other options.
+
+Overall, if no option is trivial, I feel like option `1.` is still the most pragmatic one. Option `3.` may be a better 
+design, but represents more work. It would probably be fine if all our appenders were impacted, but as only the 
+rolling-file one is, I feel like going with `3.` would be ok.
+
+### The status API
+
+In clustering mode, the workers will all have an individual status. One could have a connectivity issue with ES 
+while the other ones are green. Hitting the `/status` endpoint will reach a random (and different each time) worker, 
+meaning that it would not be possible to know the status of the cluster as a whole.
+
+We will need to add a centralized status service in the coordinator. Also, as the `/status` endpoint cannot be served 
+from the coordinator, we will also need to have the workers retrieve the global status from the coordinator to serve 
+the status endpoint.
+
+We may also want to have the `/status` endpoint display each individual worker status in addition to the 
+global status, which may be breaking change in the `/status` API response format.
+
+### PID file
+
+Without changes, each worker is going to try to write and read the same PID file. Also, this breaks the whole pid file 
+usage, as the PID stored in the file will be a arbitrary worker’s PID, instead of the coordinator (main process) PID.
+
+In clustering mode, we will need to have to coordinator handle the PID file logic, and to disable pid file handling 
+in the worker's environment service.
+
+### SavedObjects migration
+
+In the current state, all workers are going to try to perform the migration. Ideally, we would have only one process 
+perform the migration, and the other ones just wait for a ready signal. We can’t easily have the coordinator do it, 
+so we would probably have to leverage the ‘main worker’ concept here.
+
+The SO migration v2 is supposed to be resilient to concurrent attempts though, as we already support multi-instances 
+Kibana, so this can probably be considered an improvement.
+
+## Open questions and things to solve
+
+### Memory consumption
+
+In clustered mode, node options such as `max-old-space-size` will be used by all processes. 
+
+E.g using `--max-old-space-size=1024` in a 2 workers cluster would have a maximum memory usage of 3gb (1 coordinator + 2 workers). 
+
+If this something we will need to document somewhere?
```

Comment:
It'd be nice if we had the flexibility to allocate different maximum memory usage for the coordinator versus the workers. They're conceptually performing different functionality, so it'd be great to have that level of granularity.

@tylersmalley how do we currently recommend that users adjust the `max-old-space-size`?

Reply:
Currently this is defined in `config/node.options`

Reference: https://www.elastic.co/guide/en/kibana/current/production.html#memory

Reply:
> It'd be nice if we had the flexibility to allocate different maximum memory usage for the coordinator versus the workers

Not sure if we can do that using the `cluster` API. 

From the doc:, we can only pass env variables when forking

```ts
cluster.fork([env])
```

Reply:
Discussed this below:

> the solution we'd like to propose is to disable clustering if a user has `max-old-space-size` set at all, which would ensure it wouldn't be possible to hit unpredictable behavior. To opt-in for clustering, the user would update their config & remove `max-old-space-size` settings. And if no `max-old-space-size` is set, we could still consider enabling clustering by default once we are confident it is stable.

> If we still want to provide users with fine-grained control over memory-per-worker, we could update the config so users could provide a list of workers instead of a fixed number, perhaps something like:

```yaml
clustering:
  enabled: true
  workers:
    - name: foo # could perhaps use name in the logs as well for extra clarity
      max_old_space_size: 1234
    - name: bar
      max_old_space_size: 1234
```